<a href="https://colab.research.google.com/github/Alb3rtoGitHub/utilizando-chatgpt-como-asistente/blob/main/Utilizando_chatgpt_como_asistente.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <strong><font color="7b71ff">Análisis de datos con python: Utilizando ChatGPT como asistente</font></strong>


---

<img src="https://raw.githubusercontent.com/alura-cursos/ia-datascience-dataviz/main/imagens/logos/logo_branca_fundo_azul.png" alt="inserir alt">

# **Aula 1 - Conociendo los datos**

 Zoop es una gran minorista que atiende a todas las regiones de México a través de su e-commerce. Es conocida por la amplia variedad de productos que ofrece, buscando satisfacer a todo tipo de público. Posee una serie de centros de distribución (CDs) para la entrega de sus productos, la mayoría de ellos situados en el estado de Mexico. También posee una cadena de supermercados, y vamos a trabajar como científicos de datos para esta prestigiosa empresa.

### **Accesando a los datos**

In [13]:
# url = 'https://gist.githubusercontent.com/ahcamachod/24a6235f7cdf255501b5ab4f9b0fd97f/raw/fa329f6d6c1b6b00b3ba457cdc850b53210633ca/zoop_market.json'
url = 'https://gist.githubusercontent.com/ahcamachod/4b964953032a2078d67052ec8fb92e72/raw/8106d7c3f001390744ec3aa657b9847f306c9df8/zoop_supermercado.json' # sacado de Preparando el ambiente

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

Vamos a actuar como científicos de datos en Zoop.

Recibimos datos en formato de archivo JSON, estructurados en listas y diccionarios. El archivo está disponible en una URL de Github.

Necesitamos analizar esos datos para ayudar en las decisiones de la empresa. Programamos en Python y conocemos las bibliotecas de análisis de datos Numpy y Pandas. Por lo tanto, prioriza el uso de estas bibliotecas.

Igualmente, infórmanos cómo podemos leer estos datos en formato JSON y para poder visualizarlos en formato de tabla en Python.

In [14]:
import pandas as pd
import numpy as np

#Leer el archivo JSON desde la url
df = pd.read_json(url)

In [15]:
df

item_identificador tienda_identificador  ventas_totales  \
0                 FDB08               OUT018       176503.58   
1                 DRQ35               OUT049       185758.20   
2                 FDD14               OUT018       165983.94   
3                 FDY37               OUT045       314923.40   
4                 FDY59               OUT018        64782.34   
...                 ...                  ...             ...   
8545              FDY08               OUT010        28096.76   
8546              FDC41               OUT017       130163.90   
8547              NCQ53               OUT045       614533.40   
8548              FDL46               OUT017       164985.24   
8549              NCN30               OUT046        96541.00   

                                                   item  \
0     {'item_peso': 6.055, 'item_contenido_grasa': '...   
1     {'item_peso': 9.3, 'item_contenido_grasa': 'Ba...   
2     {'item_peso': 20.7, 'item_contenido_grasa': 'B...   
3     {'item_peso': 17.0, 'item_contenido_grasa': 'R...   
4     {'item_peso': 8.195, 'item_contenido_grasa': '...   
...                                                 ...   
8545  {'item_peso': 9.395, 'item_contenido_grasa': '...   
8546  {'item_peso': 15.6, 'item_contenido_grasa': 'B...   
8547  {'item_peso': 17.6, 'item_contenido_grasa': 'B...   
8548  {'item_peso': 20.35, 'item_contenido_grasa': '...   
8549  {'item_peso': 16.35, 'item_contenido_grasa': '...   

                                                 tienda  
0     {'tienda_ano_establecimiento': 2019, 'tienda_t...  
1     {'tienda_ano_establecimiento': 2009, 'tienda_t...  
2     {'tienda_ano_establecimiento': 2019, 'tienda_t...  
3     {'tienda_ano_establecimiento': 2012, 'tienda_t...  
4     {'tienda_ano_establecimiento': 2019, 'tienda_t...  
...                                                 ...  
8545  {'tienda_ano_establecimiento': 2008, 'tienda_t...  
8546  {'tienda_ano_establecimiento': 2017, 'tienda_t...  
8547  {'tienda_ano_establecimiento': 2012, 'tienda_t...  
8548  {'tienda_ano_establecimiento': 2017, 'tienda_t...  
8549  {'tienda_ano_establecimiento': 2007, 'tienda_t...  

[8550 rows x 5 columns]

In [6]:
# Mostrar las primeras filas
print(df.head())

# Ver información general del DataFrame
print(df.info())

# Estadísticas descriptivas (solo para columnas numéricas)
print(df.describe())

  item_identificador tienda_identificador  ventas_totales  \
0              FDB08               OUT018       176503.58   
1              DRQ35               OUT049       185758.20   
2              FDD14               OUT018       165983.94   
3              FDY37               OUT045       314923.40   
4              FDY59               OUT018        64782.34   

                                                item  \
0  {'item_peso': 6.055, 'item_contenido_grasa': '...   
1  {'item_peso': 9.3, 'item_contenido_grasa': 'Ba...   
2  {'item_peso': 20.7, 'item_contenido_grasa': 'B...   
3  {'item_peso': 17.0, 'item_contenido_grasa': 'R...   
4  {'item_peso': 8.195, 'item_contenido_grasa': '...   

                                              tienda  
0  {'tienda_ano_establecimiento': 2019, 'tienda_t...  
1  {'tienda_ano_establecimiento': 2009, 'tienda_t...  
2  {'tienda_ano_establecimiento': 2019, 'tienda_t...  
3  {'tienda_ano_establecimiento': 2012, 'tienda_t...  
4  {'tienda_ano_estab

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


El DataFrame "df" posee las columnas "item_identificador", "tienda_identificador", "ventas_totales", "item" y "tienda".

Las columnas "item" y "tienda" contienen diccionarios anidados dentro de cada una de las filas. ¿Cómo puedo transformar los datos de los diccionarios y columnas para hacerlos más accesibles en mi DataFrame?


In [16]:
df_items = pd.json_normalize(df["item"])
df_tiendas = pd.json_normalize(df["tienda"])

In [17]:
# Concatenar los nuevos DataFrames con el original (sin las columnas anidadas)
df = pd.concat([df.drop(columns=["item", "tienda"], axis=1), df_items, df_tiendas], axis=1)

In [19]:
df.head()

item_identificador tienda_identificador  ventas_totales  item_peso  \
0              FDB08               OUT018       176503.58      6.055   
1              DRQ35               OUT049       185758.20      9.300   
2              FDD14               OUT018       165983.94     20.700   
3              FDY37               OUT045       314923.40     17.000   
4              FDY59               OUT018        64782.34      8.195   

      item_contenido_grasa  item_visibilidad            item_tipo  \
0  Bajo Contenido de Grasa          0.031230   Frutas y Vegetales   
1  Bajo Contenido de Grasa          0.042357  Bebidas Alcohólicas   
2  Bajo Contenido de Grasa          0.170500            Enlatados   
3                  Regular          0.026623            Enlatados   
4  Bajo Contenido de Grasa          0.000000             Dulcería   

   item_precio item_cantidad_venta  tienda_ano_establecimiento tienda_tamano  \
0       160.36                None                        2019         Medio   
1       123.24                None                        2009         Medio   
2       184.13                None                        2019         Medio   
3       144.25                None                        2012          None   
4        93.15                None                        2019         Medio   

  tienda_tipo_ubicacion          tienda_tipo  
0               Nivel 3  Supermercado Tipo 2  
1               Nivel 1  Supermercado Tipo 1  
2               Nivel 3  Supermercado Tipo 2  
3               Nivel 2  Supermercado Tipo 1  
4               Nivel 3  Supermercado Tipo 2

# **Desafio: ¡hora de practicar!**

Un colega del equipo de datos de Zoop solicitó el análisis de otro conjunto de datos además del enviado por la dirección. Estos datos se refieren a las ventas totales obtenidas por 896 franquicias en el primer mes del trimestre de análisis de Zoop.
Para acceder a los datos, puedes usar el siguiente código con el enlace abajo:

url_franquicias = 'https://gist.githubusercontent.com/ahcamachod/68eede222bdf1ba0ca413d529e800e7c/raw/576b0f3856467817df0fcf041d4a0787579917b0/ventas_franquicia.json'

Sabiendo esto, vamos al desafío:

Con la ayuda de ChatGPT, busca recolectar y acceder a los datos del archivo que están en formato JSON. Es importante dejar todos los datos de manera accesible, evitando estructuras que recuerden a un archivo JSON, como diccionarios.

Diccionario de datos: ventas_franquicia.json

Columna	          Descripción
ID_tienda	        Código identificador de las franquicias
ventas_por_dia	  Promedio del total obtenido en ventas por día
ventas_tienda	    Ventas totales obtenidas en un mes
tamano_tienda	    Categoría del total obtenido en ventas ('Pequeño' y 'Medio')
cliente_por_dia	  Promedio del total de clientes por día
items_disponibles	Número de artículos diferentes disponibles

En primer lugar hacemos un pedido a ChatGPT con el siguiente Prompt:
Prompt

Estamos actuando como científicos de datos en una empresa de supermercados.

Recibimos datos en formato de archivo JSON, estructurados en listas y diccionarios. El archivo está disponible en una URL de GitHub, que está guardada en una variable llamada "url_franquicias".

Programamos en Python y tenemos conocimiento de las bibliotecas de análisis de datos Numpy y Pandas. Por lo tanto, priorice el uso de estas bibliotecas.

Por favor, infórmenos cómo podemos leer estos datos en formato JSON y cómo podemos visualizarlos en formato de tabla en Python.

Con la respuesta lo aplicamos a nuestro problema:

import pandas as pd

import requests

url_franquicias = 'https://gist.githubusercontent.com/ahcamachod/68eede222bdf1ba0ca413d529e800e7c/raw/576b0f3856467817df0fcf041d4a0787579917b0/ventas_franquicia.json'

/* Descargar el JSON desde la URL

response = requests.get(url_franquicias)
datos_json = response.json()  /* Convertir la respuesta en un diccionario de Python

/* Crear un DataFrame de Pandas
df = pd.DataFrame(datos_json)

/* Mostrar las primeras filas
print(df.head())

/* en Jupyter Notebook usa:
df.head()

In [28]:
import pandas as pd
import numpy as np
import requests

url_franquicias = 'https://gist.githubusercontent.com/ahcamachod/68eede222bdf1ba0ca413d529e800e7c/raw/576b0f3856467817df0fcf041d4a0787579917b0/ventas_franquicia.json'

# Descargar el JSON desde la URL
response = requests.get(url_franquicias)
datos_json = response.json()  # Convertir la respuesta en un diccionario de Python

# Crear un DataFrame de Pandas
df = pd.DataFrame(datos_json)

# print(df)

# Mostrar las primeras filas
print(df.head())
df.head()

  ID_tienda                                             ventas  \
0    OUT431  {'ventas_por_dia': None, 'ventas_tienda': 8222...   
1    OUT738  {'ventas_por_dia': None, 'ventas_tienda': 1032...   
2    OUT426  {'ventas_por_dia': None, 'ventas_tienda': 5548...   
3    OUT948  {'ventas_por_dia': None, 'ventas_tienda': 9641...   
4    OUT139  {'ventas_por_dia': None, 'ventas_tienda': 5680...   

                                              tienda  
0  {'tamano_tienda': 'pequeno', 'cliente_por_dia'...  
1  {'tamano_tienda': 'medio', 'cliente_por_dia': ...  
2  {'tamano_tienda': 'pequeno', 'cliente_por_dia'...  
3  {'tamano_tienda': 'Pequeno', 'cliente_por_dia'...  
4  {'tamano_tienda': 'Pequeno', 'cliente_por_dia'...  


ID_tienda                                             ventas  \
0    OUT431  {'ventas_por_dia': None, 'ventas_tienda': 8222...   
1    OUT738  {'ventas_por_dia': None, 'ventas_tienda': 1032...   
2    OUT426  {'ventas_por_dia': None, 'ventas_tienda': 5548...   
3    OUT948  {'ventas_por_dia': None, 'ventas_tienda': 9641...   
4    OUT139  {'ventas_por_dia': None, 'ventas_tienda': 5680...   

                                              tienda  
0  {'tamano_tienda': 'pequeno', 'cliente_por_dia'...  
1  {'tamano_tienda': 'medio', 'cliente_por_dia': ...  
2  {'tamano_tienda': 'pequeno', 'cliente_por_dia'...  
3  {'tamano_tienda': 'Pequeno', 'cliente_por_dia'...  
4  {'tamano_tienda': 'Pequeno', 'cliente_por_dia'...

podemos solicitar nuevamente la ayuda de ChatGPT.

Prompt:

El DataFrame "df" tiene las columnas "ID_tienda", "ventas" y "tienda".

Las columnas "ventas" y "tienda" contienen diccionarios anidados dentro de cada una de las filas. ¿Cómo puedo transformar los datos de estos diccionarios y columnas para hacerlos más accesibles en mi DataFrame?

In [31]:
df_ventas = pd.json_normalize(df['ventas'])
df_tienda = pd.json_normalize(df['tienda'])
# Concatenar los DataFrames con los datos originales (excluyendo las columnas anidadas)
df_final = pd.concat([df.drop(columns=['ventas', 'tienda']), df_ventas, df_tienda], axis=1)
# Mostrar las primeras filas del DataFrame final
print(df_final.head())
df_final.head()

  ID_tienda  ventas_por_dia  ventas_tienda tamano_tienda  cliente_por_dia  \
0    OUT431             NaN   82220.053714       pequeno            330.0   
1    OUT738             NaN  103231.728424         medio            310.0   
2    OUT426             NaN   55487.996318       pequeno            870.0   
3    OUT948             NaN   96417.999686       Pequeno           1560.0   
4    OUT139             NaN   56809.025487       Pequeno             10.0   

   items_disponibles  
0               1968  
1               2189  
2               1736  
3               1099  
4               1516  


ID_tienda  ventas_por_dia  ventas_tienda tamano_tienda  cliente_por_dia  \
0    OUT431             NaN   82220.053714       pequeno            330.0   
1    OUT738             NaN  103231.728424         medio            310.0   
2    OUT426             NaN   55487.996318       pequeno            870.0   
3    OUT948             NaN   96417.999686       Pequeno           1560.0   
4    OUT139             NaN   56809.025487       Pequeno             10.0   

   items_disponibles  
0               1968  
1               2189  
2               1736  
3               1099  
4               1516

# **Clase 2 - Exploración de los dados**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene datos referentes a las ventas de diversos productos en varias tiendas del supermercado. Se proporcionan los IDs de productos y los IDs de tiendas que comercializaron esos productos, el total recaudado en ventas, así como descripciones sobre el producto vendido y la tienda que lo vendió.

Estamos iniciando la etapa de exploración de datos, en la cual deseamos conocer los valores informados e identificar si existen valores ausentes, duplicados y otras inconsistencias.

Construye un código para ejecutar la etapa de exploración de datos usando Python, comentando el paso a paso de manera clara y objetiva.

No es necesario utilizar "print()" para mostrar el resultado de las consultas.

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene datos relacionados con las ventas durante un trimestre de un supermercado. Deseo utilizarlo para verificar el desempeño de esas ventas.

Durante la exploración del "df", identificamos algunas inconsistencias, como:

* Filas duplicadas,
* Valores nulos y
* Datos categóricos que fueron escritos de manera no estandarizada, dando la idea de que existen más categorías de las que realmente existen.

¿Qué debo hacer para garantizar que esas inconsistencias que encontré no afecten mi análisis?

# **Clase 3 - Limpieza y tratamiento de los datos**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


Durante la exploración del DataFrame "df", identificamos 27 filas duplicadas usando "df.duplicated().sum()".

El objetivo es analizar los datos en "df" para medir el desempeño de una empresa durante un trimestre de ventas.

¿Qué debemos hacer con los datos duplicados para garantizar que no haya impacto negativo en el análisis?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


Identificamos tres columnas en "df" que presentan valores nulos:

* La columna "item_cantidad_venta" tiene 8550 filas con valores nulos.
* La columna "item_peso" tiene 1469 filas con valores nulos.
* La columna "tienda_tamano" tiene 2417 filas con valores nulos.

Hubo un gran número de ventas de artículos durante el trimestre, y no hay justificación para la ausencia de datos en la columna "item_cantidad_venta", lo que la convierte en una columna prescindible.

La columna "item_peso" contiene valores del tipo float, representando el peso en gramos de los productos.

La columna "tienda_tamano" tiene tres categorías de tamaño para supermercados.

¿Cómo debo manejar estos datos nulos para garantizar que mi análisis no se vea afectado?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

La columna "item_conteudo_grasa" del DataFrame "df" presenta 5 valores únicos dados por: 'Bajo Contenido de Grasa', 'Regular', 'BTG', 'reg' y 'bajo contenido de grasa'.

Los valores únicos de la columna "item_contenido_grasa" deberían ser solo 'Bajo Contenido de Grasa' y 'Regular', pero fueron insertados sin estandarización, lo que resultó en más categorías de las deseadas.

¿Cómo puedo ajustar los datos de la columna "item_contenido_grasa" para mantener solo las dos categorías estándar?

# **Clase 4 - Visualización de datos**

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

Los datos del DataFrame "df" están limpios y tratados. Ahora, deseo iniciar la etapa de construcción de gráficos para crear visualizaciones que ofrezcan un buen soporte al análisis del desempeño de ventas de la empresa.

Las columnas disponibles en "df" son las siguientes:
- "item_identificador": Número único asignado a cada ítem
- "tienda_identificador": Nombre único de la tienda
- "ventas_totales": Valor obtenido en ventas totales
- "item_peso": Peso del ítem en gramos
- "item_contenido_grasa": Contenido de grasa del ítem
- "item_visibilidad": Valor de posicionamiento de cada ítem
- "item_tipo": Tipo de utilidad del ítem
- "item_precio": Precio unitario del ítem
- "tienda_ano_establecimiento": Año de Establecimiento de la tienda
- "tienda_tamano": Tamaño de la tienda
- "tienda_tipo_ubicacion": Tipo de ubicación de la tienda
- "tienda_tipo": Tipo de tienda

El objetivo es encontrar patrones o información que se destaquen cuando nos referimos al total obtenido en ventas especificado en la columna "ventas_totales".

Con base en la información proporcionada, ¿qué tipos de visualizaciones serían útiles para alcanzar el objetivo de analizar el desempeño de ventas?

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "tienda_tamano" con 4 categorías de tamaño para tiendas: 'Grande', 'Medio', 'Pequeño' y 'No registrado'.

Deseo crear una visualización que permita identificar el total obtenido en ventas para las cuatro categorías de tamaño de tienda y representarlas en un gráfico de barras.

Los datos en la columna "ventas_totales" necesitan ser agrupados por las categorías de "tienda_tamano".

La figura del gráfico debe seguir la identidad visual de la tienda, con el fondo de la figura y del gráfico en color "#ebebeb" y las barras en color "#e23155".

Además, es necesario mostrar el total de las ventas en millones encima de cada columna, con el texto de esta información en color "#19325c" y la estructura "MXN $ valor M".

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>


El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "tienda_tipo" con 4 categorías de tipo para tiendas: 'Supermercado Tipo 1', 'Supermercado Tipo 2', 'Mercado', 'Supermercado Tipo 3'.

Deseo crear una visualización que permita identificar el total obtenido en ventas para las cuatro categorías de tipo de tienda y representarlas en un gráfico de barras.

Los datos en la columna "ventas_totales" deben ser agrupados por las categorías de "tienda_tipo".

La figura de la gráfica debe seguir la identidad visual de la tienda, con el fondo de la figura y de la gráfica en color "#ebebeb" y las barras en color "#e23155".

Además, es necesario mostrar el total de las ventas en millones sobre cada columna, con el texto de esa información en color "#19325c" y la estructura "MXN$ valor M".

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.

> #### <strong><font color="7b71ff" size =6>Prompt:</font></strong>

El DataFrame "df" contiene la columna "ventas_totales", que informa el valor en reales obtenido en ventas de productos por varias tiendas.

Además, en "df", tenemos la columna "item_tipo" con 16 categorías de productos.

Deseo crear una visualización que permita identificar el total obtenido en ventas por categorías de productos en un gráfico de barras horizontales.

Los datos en la columna "ventas_totales" deben agruparse por las categorías de "item_tipo".

La figura de la representación debe seguir la identidad visual de la tienda, con el fondo de la figura y del gráfico en el color "#ebebeb".

Las barras deben estar en el color "#e23155".

Los valores de las barras deben estar ordenados para mantener un aspecto de embudo.

Además, es necesario mostrar el total de las ventas en millones dentro o al lado de cada barra, con el texto de esa información en el color "#19325c" y en la estructura "MXN$ valor M".

No muestres la cuadrícula en el gráfico.

Utiliza las bibliotecas de visualización de datos de Python para crear esta visualización.